src/contrib/mps/gemm.mm

```diff
@@ -0,0 +1,95 @@
+#include "../../runtime/metal/metal_common.h"
+#include <MetalPerformanceShaders/MetalPerformanceShaders.h>
+#include <dmlc/logging.h>
+#include <tvm/runtime/device_api.h>
+#include <tvm/runtime/registry.h>
+#include <tvm/runtime/util.h>
+
+namespace tvm {
+namespace contrib {
+
+using namespace runtime;
+
+TVM_REGISTER_GLOBAL("tvm.contrib.mps.matmul")
+    .set_body([](TVMArgs args, TVMRetValue *ret) {
+      DLTensor *A = args[0];
+      DLTensor *B = args[1];
+      DLTensor *C = args[2];
+      bool transa = args[3];
+      bool transb = args[4];
+      // call gemm for simple compact code.
+      CHECK_EQ(A->ndim, 2);
+      CHECK_EQ(B->ndim, 2);
+      CHECK_EQ(C->ndim, 2);
+      CHECK(C->strides == nullptr);
+      CHECK(B->strides == nullptr);
+      CHECK(A->strides == nullptr);
+      CHECK(TypeMatch(A->dtype, kDLFloat, 32));
+      CHECK(TypeMatch(B->dtype, kDLFloat, 32));
+      CHECK(TypeMatch(C->dtype, kDLFloat, 32));
+      // Get Metal device API
+      auto func = runtime::Registry::Get("device_api.metal");
+      void *dev_handle = (*func)();
+      runtime::metal::MetalWorkspace *metal_api =
+          static_cast<runtime::metal::MetalWorkspace *>(dev_handle);
+      // TODO(Check same device)
+      id<MTLDevice> dev = metal_api->GetDevice(A->ctx);
+      id<MTLCommandQueue> queue = metal_api->GetCommandQueue(A->ctx);
+      id<MTLCommandBuffer> cb = [queue commandBuffer];
+      // TODO(Check shape)
+      NSUInteger M = A->shape[0];
+      NSUInteger N = B->shape[1];
+      NSUInteger K = B->shape[0];
+      // mps a
+      MPSMatrixDescriptor *descA = [MPSMatrixDescriptor
```

Comment:
do we need to cache descriptor?

Reply:
I didn't find a setter for the descriptor in the metal api. I'm still learning the api so I may have missed a way to set it.